Copy and Paste (STClipboard v1)

darth-cheney opened the pull request:
## What ##
This PR implements a first pass at Part copy and paste functionality. To accomplish this we introduce a new `STClipboard` object on the System that handles most of the complexity. Note that it is completely separate from the OS/Browser based clipboard and copy/paste actions (see below).
  
## Implementation ##
### The `STClipboard` Object ###
`STClipboard` represents a kind of SimpleTalk-only clipboard that can be invoked directly by the System. There are two primary methods intended to be the main API:
* `copyPart(aPartObject)` -- This serializes the passed Part into a special store, along with a custom mimeType of `simpletalk/json'`. Note that these serializations, unlike the saveHTML ones, _not_ flattened.
* `pasteContentsInto(aPartObject)` -- This method will attempt to deserialize an object from the clipboard and, if available and if the target part accepts the type, paste a copy of the part into the target object.
  
### A Note on the browser Clipboard API ###
Ideally, we would make extensive use of the [Clipboard API](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API), which, with permissions granted, would allow us to intercept the copy and paste actions of the operating system and case based upon the mimeType of true clipboard contents. However, Firefox does not yet support even a functional minimal set of the behavior of this API spec.
  
The objects `STClipboard` and `STClipboardItem` have been designed to be future looking, in that they behave almost like the API will expect down the road. Once that is actually implemented in FF and major browsers, we should be able to make minor changes to have true clipboard functionality without a lot of hassle.
  
### Halo Updates ###
#### Copy and Paste Buttons ####
This PR introduces two new buttons for Halos representing copy and paste-into. For the moment these are the only ways to non-JS/console copy and paste Parts.
  
![Screenshot from 2021-03-23 14-41-44](https://user-images.githubusercontent.com/1640299/112200742-e8890e80-8be5-11eb-881d-3f5b1324a261.png)
  
For the moment the paste-into button is not particularly "smart" -- it will appear on all Parts, regardless of whether or not the STClipboard is empty or, if full, whether or not the target accepts the kind of part that will be pasted. This is something we can deal with using messages and/or events, but we need to make some decisions about how that will work.
  
#### Halo Behavior: Only One & Bug Fixes ####
This PR also fixes an issue where middle click was working for the hierarchical Halo toggling, but shift-click wasn't. Now both should have the same behavior.
  
Additionally, we've made an important change: now only one Halo can be open in the current viewport at a given time. This means that selecting open halo on a part will close all other open halos.
  
## Further Work and Decisions ##
Here are some open questions:
* Do we need to start thinking about contextual menus to make the copy/paste UI a bit cleaner?
* How should we handle the copying/pasting of Cards or whole Stacks (since there is no Halo)?
* We still have a `copyModel` command, but it is kind of dead in the water. We should think about a replacement that makes use of STClipboard to do the heavy lifting. How should these commands be structured?
* Do we want an ST-level message to be sent to all accepting Parts in the current card whenever a new STClipboardItem is added? What should the default behavior be?
  
Also feel free to note any other suggestions you have about what we can do in the future.
  
## Tests ##
There is a [preload test file](https://github.com/UnitedLexCorp/SimpleTalk/blob/eric-copy-paste/js/objects/tests/test-clipboard-with-preload.js) for these changes.
  
## Closes ##
#135

ApproximateIdentity commented:
I'm a little confused what the functionality is here. If I create two buttons and I click copy on one and then paste on the other, what is supposed to happen? Is it supposed to copy the name? The script? Everything? Right now it seems to me like nothing is happening.

darth-cheney commented:
@ApproximateIdentity Nothing happens because buttons don't accept any subparts (you can't paste a button into another button). Try making a new Area, then pasting the button into that.
  
Right now Cards don't have Halos so we need to think about how this might work for just pasting onto a card.
  
If you want to test this out manually in the browser console, use the two methods mentioned in the PR:
* `System.clipboard.copyPart(aPart)`
* `System.clipboard.pasteContentsInto(aPart)`

Both will take Part (model) objects as arguments. If the target of a pasteInto call accepts the kind of subpart that is in the STClipboard, is should create a copy and put it inside. Otherwise nothing happens.

darth-cheney commented:
@ApproximateIdentity @dkrasner I've rebased and also added some new functionality, as follows:
* the `paste` command will attempt to paste the contents of the STClipboard into the part that sends it, ie:
  `tell first area of this card to paste`
* the `copy` command will similarly copy the part that sends it into the STClipboard:
   `tell first button of current card to copy`
  
I have decided not to do a contextual menu just yet -- we are going to want the field integration and a couple of other things before we can do it properly. I'm thinking what we have here is enough to get started. Thoughts?

dkrasner commented:
@darth-cheney this is great

I added a button "Paste here" with the following script
```
on click
   tell current card to paste
end click
```

and then went around copying and pasting into the current card whatever i wanted. 

I imagine that we can combine this with the `onOpenX` command and create stacks/cards by copying from catalogues

```
on openCard
     tell window "Toolbox" of stack "Catalogue" to copy
     tell current card to paste
     tell window "Toolbox" of current card to set "top" to 100
end openCard
```

If we sort out how to deal with implicit 'it' we could write things like `tell it to set...` making pretty readable, flexible scripts. 

there is one test failing
![image](https://user-images.githubusercontent.com/1154326/112671341-580b3200-8e62-11eb-87e2-3abd86bac175.png)

which looks like some minor props issue. Otherwise merge it in!

> * Do we need to start thinking about contextual menus to make the copy/paste UI a bit cleaner?
yes but lets see if we can build them out of ST parts eventually
> 
> * How should we handle the copying/pasting of Cards or whole Stacks (since there is no Halo)?
one way to do it is with a custom button like the one i made above. Another way is in your navigator work. Lets see how that pans out and feels. 
At least for the moment it's really easy to make a button on a stack to paste into the current, or any, card.
> 
> * We still have a `copyModel` command, but it is kind of dead in the water. We should think about a replacement that makes use of STClipboard to do the heavy lifting. How should these commands be structured?
> 
i think we should get rid of this and see where copying takes us. 
> * Do we want an ST-level message to be sent to all accepting Parts in the current card whenever a new STClipboardItem is added? What should the default behavior be?
yes I think this is a good idea. Maybe we can have a (private) part prop like "canAcceptClipboard" which would be set on this `newCliopboardItem`   That way halo can simply check if the current part accepts what's in clipboard and shows/hides the paste icon
